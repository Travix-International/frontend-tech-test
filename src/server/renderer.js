import React from 'react';
import ReactServer from 'react-dom/server';
import { StaticRouter as Router, matchPath } from 'react-router';
import { Provider } from 'react-redux';
import { Helmet } from 'react-helmet';
import { /* get as _get, */ isFunction as _isFunction } from 'lodash';
import renderFullPage from './layout';
import configureStore from '../configureStore';
import { entryRouteComponentMap, getEntryAndRoute, entryReducerMap } from './entryAndRoute';

function applyInitStore(req) {
  const entryRouteInfo = getEntryAndRoute(req.path);
  const currentEntry = entryRouteInfo.entry;
  const rootReducer = entryReducerMap[currentEntry];
  const store = configureStore({}, rootReducer);
  req.reduxStore = store;
  req.entryRouteInfo = entryRouteInfo;
  // next();
}

function applyRouteCheckResult(req) {
  const { path, reduxStore: store, query, entryRouteInfo } = req;
  const promises = [];
  const currentEntry = entryRouteInfo.entry;
  const route = entryRouteInfo.route;
  const match = matchPath(path, route);
  const { loadData } = route;

  // set currentEntry to req
  req.currentEntry = currentEntry;

  // it's possible no loadData set
  if (_isFunction(loadData)) {
    let action = loadData(match, query);
    if (action) {
      if (_isFunction(action) || action.type) {
        action = store.dispatch(action);
        if (action.then) {
          promises.push(action);
        }
      }
    }
  }

  return Promise.all(promises);
}

function responsePage(req, res, clientStats) {
  const { url, reduxStore: store, currentEntry } = req;
  const routerContext = {};
  const reduxStateString = JSON.stringify(store.getState()).replace(/</g, '\\u003c');

  const RouteComponent = entryRouteComponentMap[currentEntry];

  const content = (
    <Provider store={store}>
      <Router context={routerContext} location={url}>
        <RouteComponent />
      </Router>
    </Provider>
  );

  // render to sting to get helmet setting
  const contentString = ReactServer.renderToString(content);
  const head = Helmet.renderStatic();
  const htmlString = renderFullPage(contentString, reduxStateString, head, currentEntry, clientStats);

  res.send(htmlString);
}

export default function renderer({ clientStats }) {
  return (req, res) => {
    applyInitStore(req);
    return applyRouteCheckResult(req).then(() => {
      return responsePage(req, res, clientStats);
    });
  };
}
