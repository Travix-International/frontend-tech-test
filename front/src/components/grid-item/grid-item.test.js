import React from 'react'
import renderer from 'react-test-renderer'

import GridItem from '.'

describe('GridItemComponent', () => {
  it('Snapshot', async () => {
    const tree = renderer
      .create(<GridItem
        title={'title'}
        description={'description'}
        handleRemove={()=>{}}
        handleEdit={()=>{}}
      />)
      .toJSON()
    expect(tree).toMatchSnapshot()
  })
})
