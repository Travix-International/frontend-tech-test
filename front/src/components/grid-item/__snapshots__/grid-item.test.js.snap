// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GridItemComponent Snapshot 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c0:hover {
  background: rgba(0,0,0,.05);
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 0 1rem;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  margin-left: .5rem;
}

.c3 {
  color: #707070;
  font-size: 1.6rem;
  font-weight: 400;
  display: inline-block;
  margin-top: .7rem;
  margin-bottom: .5rem;
  vertical-align: top;
}

.c4 {
  color: #707070;
  font-size: 1.2rem;
  margin: 0 0 1rem;
}

.c5 {
  width: 2rem;
  height: 2rem;
  fill: #707070;
  margin: 0 1rem;
  cursor: pointer;
}

.c6 {
  width: 2rem;
  height: 2rem;
  fill: #707070;
  margin: 0 1rem;
  cursor: pointer;
}

<li
  className="c0"
>
  <div
    className="c1"
  >
    <div
      className="c2"
    >
      <h1
        className="c3"
      >
        title
      </h1>
      <p
        className="c4"
      >
        description
      </p>
    </div>
  </div>
  <div>
    <test-file-stub
      className="c5"
      onClick={[Function]}
    />
    <test-file-stub
      className="c6"
      onClick={[Function]}
    />
  </div>
</li>
`;
